File: 2024/Day2/2.py
def all_increasing_or_decreasing(arr):
    """Check if all levels are increasing or decreasing."""
    return all(arr[i] <= arr[i + 1] for i in range(len(arr) - 1)) or \
           all(arr[i] >= arr[i + 1] for i in range(len(arr) - 1))

def is_differ_at_least_1_and_at_most_3(arr):
    """Check if all differences are between 1 and 3 inclusive."""
    return all(1 <= abs(arr[i] - arr[i - 1]) <= 3 for i in range(1, len(arr)))

def is_safe(arr):
    """Check if the report is safe without any modifications."""
    return all_increasing_or_decreasing(arr) and is_differ_at_least_1_and_at_most_3(arr)

def can_be_safe_by_removing_one(arr):
    """Check if the report can be made safe by removing a single level."""
    for i in range(len(arr)):
        modified_arr = arr[:i] + arr[i + 1:]  # Remove the i-th level
        if is_safe(modified_arr):
            return True
    return False

# Reading the input data
with open("input.txt", "r") as file:
    content = file.readlines()

safe_reports_count = 0

for line in content:
    levels = list(map(int, line.strip().split()))
    if is_safe(levels) or can_be_safe_by_removing_one(levels):
        safe_reports_count += 1

print(safe_reports_count)

File: 2024/Day2/2.swift
import Foundation

// Check if all levels are increasing or decreasing
func allIncreasingOrDecreasing(_ arr: [Int]) -> Bool {
    let increasing = zip(arr, arr.dropFirst()).allSatisfy { $0 <= $1 }
    let decreasing = zip(arr, arr.dropFirst()).allSatisfy { $0 >= $1 }
    return increasing || decreasing
}

// Check if all differences are between 1 and 3 inclusive
func isDifferAtLeast1AndAtMost3(_ arr: [Int]) -> Bool {
    return zip(arr, arr.dropFirst()).allSatisfy { abs($0 - $1) >= 1 && abs($0 - $1) <= 3 }
}

// Check if the report is safe without any modifications
func isSafe(_ arr: [Int]) -> Bool {
    return allIncreasingOrDecreasing(arr) && isDifferAtLeast1AndAtMost3(arr)
}

// Check if the report can be made safe by removing a single level
func canBeSafeByRemovingOne(_ arr: [Int]) -> Bool {
    for i in 0..<arr.count {
        var modifiedArr = arr
        modifiedArr.remove(at: i)
        if isSafe(modifiedArr) {
            return true
        }
    }
    return false
}

// Reading the input file
let filePath = "input.txt"
var safeReportsCount = 0

do {
    let content = try String(contentsOfFile: filePath, encoding: .utf8)
    let lines = content.split(separator: "\n")

    for line in lines {
        let levels = line.split(separator: " ").compactMap { Int($0) }
        if isSafe(levels) || canBeSafeByRemovingOne(levels) {
            safeReportsCount += 1
        }
    }

    print("Safe Reports Count: \(safeReportsCount)")

} catch {
    print("Error reading file: \(error)")
}
